prow/cmd/webhook-server/helpers.go

```diff
@@ -246,37 +254,183 @@ func createValidatingWebhookConfig(ctx context.Context, caPem string, client ctr
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to create mutating webhook config: %w", err)
+	}
+
+	return nil
+}
+
+func patchMutatingWebhookConfig(ctx context.Context, caPem string, client ctrlruntimeclient.Client) error {
+	key := types.NamespacedName{
+		Namespace: defaultNamespace,
+		Name:      prowJobMutatingWebhookName,
+	}
+
+	patchOptions := &ctrlruntimeclient.PatchOptions{
+		FieldManager: "webhook-server",
+	}
+	var mutatingWebhookConfig admregistration.MutatingWebhookConfiguration
+	if err := client.Get(ctx, key, &mutatingWebhookConfig); err != nil {
+		return fmt.Errorf("failed to get mutating webhook config: %w", err)
+	}
+	oldMutatingWebhook := mutatingWebhookConfig.DeepCopy()
+	mutatingWebhookConfig.Webhooks[0].ClientConfig.CABundle = []byte(caPem)
+	if err := client.Patch(ctx, &mutatingWebhookConfig, ctrlruntimeclient.MergeFrom(oldMutatingWebhook), patchOptions); err != nil {
+		return fmt.Errorf("failed to patch mutating webhook config: %w", err)
+	}
+	return nil
+}
+
+func checkWebhooksExist(ctx context.Context, client ctrlruntimeclient.Client) (string, string, bool, error) {
+	var mutatingExists bool
+	var validatingExists bool
+	var mutatingWebhookConfig admregistration.MutatingWebhookConfiguration
+	var validatingWebhookConfig admregistration.ValidatingWebhookConfiguration
+	mutatingKey := types.NamespacedName{
+		Namespace: defaultNamespace,
+		Name:      prowJobMutatingWebhookName,
+	}
+	validatingKey := types.NamespacedName{
+		Namespace: defaultNamespace,
+		Name:      prowJobValidatingWebhookName,
+	}
+
+	err := client.Get(ctx, mutatingKey, &mutatingWebhookConfig)
+	if err != nil && strings.Contains(err.Error(), "not found") {
+		return "", "", false, nil
+	} else if err != nil {
+		return "", "", false, fmt.Errorf("error getting mutating webhook config %v", err)
+	}
+	mutatingExists = true
+
+	err = client.Get(ctx, validatingKey, &validatingWebhookConfig)
+	if err != nil && strings.Contains(err.Error(), "not found") {
+		return "", "", false, nil
+	} else if err != nil {
+		return "", "", false, fmt.Errorf("error getting validating webhook config %v", err)
+	}
+	validatingExists = true
+
+	if mutatingExists && validatingExists {
+		return string(mutatingWebhookConfig.Webhooks[0].ClientConfig.CABundle), string(validatingWebhookConfig.Webhooks[0].ClientConfig.CABundle), true, nil
+	}
+
+	return "", "", false, nil
+}
+
+func reconcileWebhooks(ctx context.Context, caPem string, cl ctrlruntimeclient.Client) error {
+	mutatingCAPem, validatingCAPem, exist, err := checkWebhooksExist(ctx, cl)
+	if err != nil {
+		return err
+	}
+	if exist && (validatingCAPem != caPem || mutatingCAPem != caPem) {
+		if err := patchValidatingWebhookConfig(ctx, caPem, cl); err != nil {
+			return fmt.Errorf("unable to patch ValidatingWebhookConfig %v", err)
+		}
+		if err := patchMutatingWebhookConfig(ctx, caPem, cl); err != nil {
+			return fmt.Errorf("unable to patch MutatingWebhookConfig %v", err)
+		}
+	} else if !exist {
```

Comment:
I think there is a bug here, if validatingwebhook exist and mutatingwebhook doesn't, it'll fail at line 423 and return error

Reply:
if one does not exist, it is treated as neither existing. so it should be fine

Reply:
not entirely, if anyone exists then the `create` call would return error right? we shouldn't let the binary fail in this case

Reply:
The create code handles this situation by patching in that case.

Reply:
I see, it's not very intuitive to imagine that a function called `createValidatingWebhookConfig` would also do patching, I think it might be better to rename it to `ensureValidatingWebhookConfig`